### ISSUE TITLE
Use case: jsPDF

### ISSUE THREAD

**Issue by whitetigle** [fable-compiler/ts2fable#148]
Hi! I just wanted to share this experience with you. I wanted to use jsPDF yesterday in my project and I found the following "errors":

```typescript
// Type definitions for jsPDF v1.1.135
// Project: https://github.com/MrRio/jsPDF
// Definitions by: Amber Schühmacher <https://github.com/amberjs>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

declare module 'jspdf' {
    class jsPDF {
        constructor(orientation?:any,
                    unit?:string,
                    format?:string,
                    compressPdf?:number);

        CapJoinStyles:any;
        version:string;

        internal: {
            'pdfEscape'(text:string, flags:any): any;
            'getStyle'(style:string) : any;
            'getFont'(): any;
            'getFontSize'():number;
            'getLineHeight'():number;
            'write'(string1:string):any;
            'getCoordinateString'(value:number):number;
            'getVerticalCoordinateString'(value:number):number;
            'collections':any;
            'newObject'():number;
            'newAdditionalObject'():any;
            'newObjectDeferred'():number;
            'newObjectDeferredBegin'(oid:number):void;
            'putStream'(str:string):void;
            'events':any;
            'scaleFactor':number;
            'pageSize': {
                width:number;
                height:number;
            };
            'output'(type:any, options:any):any;
            'getNumberOfPages'():number;
            'pages':number[];
            'out'(string:string):void;
            'f2'(number:number):number;
            'getPageInfo'(pageNumberOneBased:number):any;
            'getCurrentPageInfo'():any;
        };
        addPage(sizes?: number[]):jsPDF;
        setPage(n:number):jsPDF;
        insertPage(beforePage:number):jsPDF;
        movePage(targetPage:number, beforePage:number):jsPDF;
        deletePage(n:number):jsPDF;
        setDisplayMode(zoom?:string, layout?:string, pmode?:string):jsPDF;
        text(text:any, x:any, y:any, flags?:any, angle?:any, align?:any):jsPDF;
        lstext(text:string, x:number, y:number, spacing:number):jsPDF;
        line(x1:number, y1:number, x2:number, y2:number):any;
        clip():void;
        lines(lines:any, x:any, y:any, scale?:any, style?:string, closed?:boolean):jsPDF;
        rect(x:number, y:number, w:number, h:number, style?:string):jsPDF;
        triangle(x1:number, y1:number, x2:number, y2:number, x3:number, y3:number, style:string):jsPDF;
        roundedRect(x:number, y:number, w:number, h:number, rx:number, ry:number, style:string):jsPDF;
        ellipse(x:number, y:number, rx:number, ry:number, style?:string):jsPDF;
        circle(x:number, y:number, r:number, style:string):jsPDF;
        setProperties(properties:any):jsPDF;
        setFontSize(size:number):jsPDF;
        setFont(fontName?:string, fontStyle?:string):jsPDF;
        setFontStyle(style:string):jsPDF;
        setFontType(style:string):jsPDF;
        getFontList():any;
        addFont(postScriptName:string, fontName:string, fontStyle:string):string;
        setLineWidth(width:number):jsPDF;
        setDrawColor(ch1:number|string, ch2?:number, ch3?:number, ch4?:number):jsPDF;
        setFillColor(ch1:number|string, ch2?:number, ch3?:number, ch4?:number):jsPDF;
        setTextColor(r?:number, g?:number, b?:number):jsPDF;
        setLineCap(style:string|number):jsPDF;
        setLineJoin(style:string|number):jsPDF;
        output(type?:string, options?:any):any;
        save(filename:string):jsPDF;

        /**
         * jsPDF plugins below:
         *
         *  - AddHTML
         *  - AddImage
         *  - Annotations
         *  - AutoPrint
         *  - Canvas
         *  - Cell
         *  - Context2D
         *  - FromHTML
         *  - JavaScript
         *  - PNG
         *  - split_text_to_size
         *  - SVG
         *  - total_pages
         */

        // jsPDF plugin: addHTML
        addHTML(element:any, x:number, y:number, options:any, callback:Function):jsPDF;
        addHTML(element:any, callback:Function):jsPDF;

        // jsPDF plugin: addImage
        color_spaces:any;
        decode:any;
        image_compression:any;

        sHashCode(str:string):any;
        isString(object:any):boolean;
        extractInfoFromBase64DataURI(dataURI:string):any[];
        supportsArrayBuffer():boolean;
        isArrayBuffer(object:any):boolean;
        isArrayBufferView(object:any):boolean;
        binaryStringToUint8Array(binary_string:string):Uint8Array;
        arrayBufferToBinaryString(buffer:any):string;
        arrayBufferToBase64(arrayBuffer:ArrayBuffer):string;
        createImageInfo(data:any, wd:any, ht:any, cs:any, bpc:any, imageIndex:number, alias:any, f?:any, dp?:any, trns?:any, pal?:any, smask?:any):any;
        addImage(imageData?:any, format?:any, x?:number, y?:number, w?:number, h?:number, alias?:any, compression?:any, rotation?:any):jsPDF;
        processJPEG(data:any, index:number, alias:any, compression?:any, dataAsBinaryString?:string):any;
        processJPG():any;

        // jsPDF plugin: Annotations
        annotationPlugin:any;
        createAnnotation(options:any):void;
        link(x:number, y:number, w:number, h:number, options:any):void;
        textWithLink(text:string, x:number, y:number, options:any):number;
        getTextWidth(text:string):number;
        getLineHeight():number;

        // jsPDF plugin: AutoPrint
        autoPrint():jsPDF;

        // jsPDF plugin: Canvas
        canvas: {
            getContext():any;
            style:any;
        };

        // jsPDF plugin: Cell
        setHeaderFunction(func:Function):void;
        getTextDimensions(txt:string):any;
        cellAddPage():void;
        cellInitialize():void;
        cell(x:number, y:number, w:number, h:number, txt:string, ln:number, align:string):jsPDF;
        arrayMax(array:any[], comparisonFn?:Function):number;
        table(x:number, y:number, data:any, headers:string[], config:any):jsPDF;
        calculateLineHeight(headerNames:string[], columnWidths:number[], model:any[]):number;
        setTableHeaderRow(config:any[]):void;
        printHeaderRow(lineNumber:number, new_page?:boolean):void;

        // jsPDF plugin: Context2D
        context2d: {
            pageWrapXEnabled: boolean;
            pageWrapYEnabled: boolean;
            pageWrapX: number;
            pageWrapY: number;
            f2(number:number):number;
            fillRect(x:number, y:number, w:number, h:number):void;
            strokeRect(x:number, y:number, w:number, h:number):void;
            clearRect(x:number, y:number, w:number, h:number):void;
            save():void;
            restore():void;
            beginPath():void;
            closePath():void;
            setFillStyle(style:string):void;
            setStrokeStyle(style:string):void;
            fillText(text:string|string[], x:number, y:number, maxWidth:number):void;
            strokeText(text:string|string[], x:number, y:number, maxWidth:number):void;
            setFont(font:string):void;
            setTextBaseline(baseline:string):void;
            getTextBaseline():string;
            setLineWidth(width:number):void;
            setLineCap(style:string):void;
            setLineJoin(style:string):void;
            moveTo(x:number, y:number):void;
            lastBreak: number;
            pageBreaks: any[];
            lineTo(x:number, y:number):void;
            bezierCurveTo(x1:number, y1:number, x2:number, y2:number, x:number, y:number):void;
            quadraticCurveTo(x1:number, y1:number, x:number, y:number):void;
            arc(x:number, y:number, radius:number, startAngle:number, endAngle:number, anticlockwise:any):void;
            drawImage(img:string, x:number, y:number, w:number, h:number, x2?:number, y2?:number, w2?:number, h2?:number):void;
            stroke():void;
            fill():void;
            translate(x:number, y:number):void;
            measureText(text:string):number;
        };

        // jsPDF plugin: fromHTML
        fromHTML(HTML:string | HTMLElement, x:number, y:number, settings?:any, callback?:Function, margins?:any):jsPDF;

        // jsPDF plugin: JavaScript
        addJS(txt:string):jsPDF;

        // jsPDF plugin: PNG
        processPNG(imageData:any, imageIndex:number, alias:string, compression:any, dataAsBinaryString:string):any;

        // jsPDF plugin: split_text_to_size
        getCharWidthsArray(text:string, options?:any):any[];
        getStringUnitWidth(text:string, options?:any):number;
        splitTextToSize(text:string, maxlen:number, options?:any):any;

        // jsPDF plugin: SVG
        addSVG(svgtext:string, x:number, y:number, w?:number, h?:number):jsPDF;

        // jsPDF plugin: total_pages
        putTotalPages(pageExpression:string):jsPDF;
    }
    
    namespace jsPDF {}

    export = jsPDF;
}

```

### 1. empty module
At the end, `namespace jsPDF {}` becomes a `module jsPDF =` in F# with nothing inside, which raises an error.

### 2. resolve type
        `fromHTML(HTML:string | HTMLElement, x:number, y:number, settings?:any, callback?:Function, margins?:any):jsPDF;`

is rightly converted  to 

`member __.fromHTML(HTML: U2<string, HTMLElement>, x: float, y: float, ?settings: obj, ?callback: Function, ?margins: obj): jsPDF = jsNative
`        
but maybe, it would be cool to detect that we need to open Fable.Import.Browser to resolve the HTMLElement type. 

### 3. missing Interface?
```ts
internal: {
            'pdfEscape'(text:string, flags:any): any;
            'getStyle'(style:string) : any;
            'getFont'(): any;
            'getFontSize'():number;
            etc...
```

is converted to 

`member __.``internal`` with get(): obj = jsNative and set(v: obj): unit = jsNative`

Same case for 
```ts
context2d: {
            pageWrapXEnabled: boolean;
            pageWrapYEnabled: boolean;
            pageWrapX: number;
            pageWrapY: number;
```

So maybe we are missing some interface there?

### 4. import rule
Here, I'm clearly not a specialist and it may not be a prblem with the correct import statement. If that's the case it would be uber cool to have a new tutorial on how to import node modules from fable with several use cases. 
Because, even if we do have two great docs (one from @alfonsogarciacaro and one from @Zaid-Ajaj ) maybe it would help close the gap from JS newcomers.

Anyway in the generated wrappers, we have 
```f#
module jspdf =
    type [<AllowNullLiteral>] [<Import("jsPDF","jspdf")>] jsPDF(?orientation: obj, ?unit: string, ?format: string, ?compressPdf: float) =
```

but in order to import the file in my code I had to replace it to:
```f#
module jspdf =
    type [<AllowNullLiteral>] [<Import("*","jspdf")>] jsPDF(?orientation: obj, ?unit: string, ?format: string, ?compressPdf: float) =
```
So that I could call it with 

`let jsPDF : jsPDF = importAll  "../node_modules/jspdf"`

But then I had to use the `new `keyword to actually instantiate jsPDF: 
`let doc = new jsPDF("a4.pdf")`

Maybe replacing jsPDF with * was not useful and there was another way to do that?

### conclusion
I will send you other reports if you wish since I use many libs. 
Last bu not least, with latest ts2fable it's been easier to generate wrappers. So I wish to thank you for all the energy you put into this project! It's helping people every day!

Thanks!

**Comment by whitetigle** [comment]
Oh well it seems my version is outdated. And I was missing some points completely. I will check that and report here.

**Comment by whitetigle** [comment]
Well after getting the same version via `yarn global ts2fable@next` (ts2fable@0.6.0-build.165) I got the same results. (which if I remember well, I did just last week)

BTW how can I know the version of ts2fable I'm using? Is there a `--version` arg?

**Comment by whitetigle** [comment]
ok it seems that using `yarn global ts2fable@next` I do get the 0.6.0 build 

```
success Installed "ts2fable@0.6.0-build.165" with binaries:
      - ts2fable

```
but then when I do yarn info on the package, it seems I'm using 0.5.2:

```
{ name: 'ts2fable',
  description: 'TypeScript definition files parser for fable-compiler',
  'dist-tags':
   { latest: '0.5.2',
     next: '0.6.0-build.165' },
```
and
```
readmeFilename: 'README.md',
  users: { ctaggart: true },
  version: '0.5.2',
  main: './dist/ts2fable.js',
  bin: { ts2fable: './dist/ts2fable.js' },
  files: [ 'dist/' ],
```

How can i know which version I'm relaly using?

**Comment by whitetigle** [comment]
Ok. So in order to get sure to not waste time on regressions issues. 
I got the sources. Compiled and tested. And now I get totally different results. (just awesome 👏 ) 

So it seems that all this time I was using a wrong version of ts2fable (0.5.2). I was misleaded by yarn messages. 

I mean: 
```

success Installed "ts2fable@0.6.0-build.165" with binaries:
      - ts2fable
```

was just telling me that ts2fable was installed, but not that 0.6.0 was installed.
So either I've got something broken on my system somewhere either there's another way to force a version. Maybe it's the window version of yarn?

Anyway, Sorry for this morning mess. I just lost time for nothing. Hope I did not make you lose time either.

Have a nice day!
